src/EditorFeatures/Core/Implementation/EditAndContinue/EditAndContinueDiagnosticAnalyzer.cs

```diff
@@ -47,22 +52,70 @@ public override Task<ImmutableArray<Diagnostic>> AnalyzeSemanticsAsync(Document
                 return SpecializedTasks.EmptyImmutableArray<Diagnostic>();
             }
 
-            return AnalyzeSemanticsImplAsync(document, cancellationToken);
+            return AnalyzeSemanticsImplAsync(workspace, document, cancellationToken);
+        }
+
+        // Copied from
+        // https://github.com/dotnet/sdk/blob/main/src/RazorSdk/SourceGenerators/RazorSourceGenerator.Helpers.cs#L32
+        private static string GetIdentifierFromPath(string filePath)
+        {
+            var builder = new StringBuilder(filePath.Length);
+
+            for (var i = 0; i < filePath.Length; i++)
+            {
+                switch (filePath[i])
+                {
+                    case ':' or '\\' or '/':
+                    case char ch when !char.IsLetterOrDigit(ch):
+                        builder.Append('_');
+                        break;
+                    default:
+                        builder.Append(filePath[i]);
+                        break;
+                }
```

Comment:
I realise this is a copy and paste, but this code seems odd. `:`, `\` and `/` all return false for `char.IsLetterOrDigit` anyway. Isn't this whole switch just `builder.Append(char.IsLetterOrDigit(filePath[i]) ? filePath[i] : '_';`?

Is there some perf benefit to a switch here? Or does it need to be exactly the same as Razor for the benefit of their parser or something?

Reply:
@pranavkm